BIP138: Compact Encryption Scheme for Non-seed Wallet Data

pythcoiner opened the pull request:
This is a bip for encrypted backup, an encryption scheme for bitcoin wallet related metadata.

Mailing list post: https://groups.google.com/g/bitcoindev/c/5NgJbpVDgEc

pythcoiner committed:
bip encrypted_backup

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+This BIP defines a compact encryption scheme for **wallet descriptors** (BIP-0380),
+**wallet policies** (BIP-0388), **labels** (BIP-0329), and
+**wallet backup metadata** (json). The encrypted output—called a
+*bitcoin encrypted backup* (BEB), enables users to outsource long‑term storage to
```
This scheme is generic enough that I don't think it needs the word "bitcoin" in it. I also don't think we need to pick a "marketing name", something more boring like "BIP_XXX encrypted backup" should be fine.

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+* Let `p_1`, `p_2`, .., `p_n` be the public keys in the descriptor/wallet policy, in
+increasing lexicographical order.  
+* let `secret` = sha256("BEB_DECRYPTION_SECRET" ‖ `p1` ‖ `p2` ‖ ... ‖ `pn`)  
+* let `si` = sha256("BEB_INDIVIDUAL_SECRET" ‖ `pi`)  
```
If it's possible to use subscript in markdown then I would prefer to keep the original notation with small i. Otherwise let's use an underscore, i.e. `s_i`

pythcoiner replied:
done

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+
+#### Content
+
+`CONTENT`: 1-byte unsigned integer identifying what has been encrypted.
```
I would prefer to keep this secret _and_ have multiple pieces of content in a single backup. So my suggestion is to move this inside the encrypted payload and then allow for multiple content entries.

pythcoiner replied:
Yeah this BIP draft differ a bit from the initial post in the fact it intentionally add some formating convention

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+| 0x01   | BIP-0380 Descriptor (string)           |
+| 0x02   | BIP-0388 Wallet policy (string)        |
+| 0x03   | BIP-0329 Labels (JSONL)                |
+| 0x04   | Wallet backup (JSON)                   |
```
Maybe point out that unlisted types should still be decrypted and that new types may be added to the BIP.

That said, a more generic approach would be to allow a bit more space for the type. E.g. a ~4 digit number that must represent a BIP number. Or a short string like "BIP-0329". Both avoid the need to maintain a central register.

Also, if you allow multiple pieces of content in a single encrypted blob, it's useful to have a padding type.

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+The version field enables possible future enhancements:
+
+- Additional encryption algorithms  
+- Support for threshold-based decryption
```
- hide the number of participants
- bech32m export (convenient for text based storage like an email to yourself, as well as for printing - if you're patient)

pythcoiner replied:
> hide the number of participants

hum, the number of participant can also be hidden (derivation path are optionals) and one can also add fake ones, how do you think we can enhance this further?

Sjors replied:
It's currently revealed by the number of s_i shares, so it would be a more complicated overhaul. I don't think it's urgent.

pythcoiner replied:
added

Sjors commented:
> **Open questions**

> - **Deterministic nonce:** Currently the nonce is generated randomly. Is it safe to produce a deterministic nonce, e.g. `hash("NONCE" || plaintext || key_1 || … || key_n)`, or are there known security concerns with this approach?  

In general nonce reuse is unsafe because if you make multiple backups over time, e.g. as you add more transaction labels, you would be reusing the nonce with different message. By including the `plaintext` in the nonce, you do address that concern.

However it still seems unwise to mess with cryptographic standards. It doesn't seem worth the risk for saving 32 bytes on something that's going to be at least a few hundred bytes for a typical multisig.

shocknet-justin commented:
Concept ACK, seems adjacent to how some lightning tools enable users to recover SCB's with just their seed to identify and decrypt the backup. Makes sense for descriptors to have something similar.

pythcoiner commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+
+* can be **safely stored in multiple places**, including untrusted on-line services,  
+* can be **decrypt only by intended holders** of specified public keys,  
+* and keeps both **keys and plaintext** hidden from any party who lacks the
```
I'll drop this one, it's redunndant with the previous one

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,268 @@
+
+* Let `p_1`, `p_2`, .., `p_n` be the public keys in the descriptor/wallet policy, in
+increasing lexicographical order.  
+* let `secret` = sha256("BEB_DECRYPTION_SECRET" ‖ `p1` ‖ `p2` ‖ ... ‖ `pn`)  
```
Liana could switch to the BIP number once assigned and when reading also support BEP. That also gives you the flexibility to make breaking changes to other parts of this spec without having to bump the version number.

pythcoiner replied:
Agree, we are in control of what happen in Liana, but my concern is more about other implementations that will not be able to decrypt early version generated w/ Liana...
Btw, we already plan to ship w/ version 0, that's why this BIP draft define version 1...

pythcoiner commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,231 @@
+
+This BIP defines a compact encryption scheme for **wallet descriptors** (BIP-0380),
+**wallet policies** (BIP-0388), **labels** (BIP-0329), and
+**wallet backup metadata** (json). The payload must not contains any private key material.
```
done

pythcoiner commented:
> Hi @pythcoiner,
> 
> By coincidence, two weeks ago I started working on a proposal for a "Standard Encrypted Wallet Payload" to be placed inside an "Encrypted Envelope". The "Wallet Payload" contains descriptors and metadata but can also act as a full wallet backup including transactions, UTXOs and addresses. The proposal is very much a work in progress.

Hi @KeysSoze, this work seems more related/parallel to the [`wallet_backup`](https://github.com/pythcoiner/wallet_backup) specs I've work few month ago.

But I've adopted a slightly different approach by simply using JSON.

FYI we already implemented this wallet backup format in Liana wallet and I plan to work on a BIP proposal relatively soon.

abrahem79 reviewed (commented):
bip-encrypted-backup.md

Sjors commented:
The PR is still draft. Are there things you're still working on?

What's the `bip.diff` file about? Accidentally committed?

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,306 @@
+   Cloud encryption doesn't help against subpoenas or provider breaches, and each
+   copy increases attack surface.  
+
+These constraints lead to an acute need for an **encrypted**, and
```
"acute"? It's been years :-)

> This BIP therefore proposes an **encrypted**, and **compact** backup format that:

Sjors commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,360 @@
+
+The format uses CHACHA20_POLY1305 (RFC 8439) as the default encryption algorithm,
+with a 96-bit random nonce and a 128-bit authentication tag to provide confidentiality
+and integrity. AES_256_GCM is also supported as an alternative.[^chacha-default]
```
I'm not sure it's a good idea to support two formats. On the encryption side there's no extra burden, since implementers can just pick their favourite. But on the decryption side you kinda have no choice but to support both.

My preference would be to drop AES_256_GCM entirely. Alternatively to make it a second class citizen, with phrasing along the lines of:

> Implementations MUST support CHACHA20_POLY1305 decryption and MAY also support the alternative AES_256_GCM scheme for decryption . 

> Implementations SHOULD use CHACHA20_POLY1305 encryption, but if implementation is problematic they MAY use AES_256_GCM instead.

craigraw commented:
This proposal makes multiple references to BIP 380 as defining wallet descriptors. This is incorrect - BIP 380 defines output script descriptors. Since we may have a need to define wallet descriptors as specific format in future (including, for example, a wallet birthday) I believe these references should be corrected.

Sjors commented:
Thanks for the updates. I plan to update my branch as well, but might be some time.

Regarding varint, it might slightly simplify a function like [DecodeContent](https://github.com/Sjors/bitcoin/pull/109/changes#diff-4ce9a1b01f4f280761f7d6cff39806d8af106bb778d30201c1527f99336c299dR288-R332), but it's not a big deal.

pythcoiner commented:
> Thanks for the updates. I plan to update my branch as well, but might be some time.
> 
> Regarding varint, it might slightly simplify a function like [DecodeContent](https://github.com/Sjors/bitcoin/pull/109/changes#diff-4ce9a1b01f4f280761f7d6cff39806d8af106bb778d30201c1527f99336c299dR288-R332), but it's not a big deal.

I not yet take time to look at your implem, but I plan to do so soon as I'm quite done with implem update on the rust implem and want to try tests vectors on both implem before finalizing the spec

Sjors commented:
My implementation should be taken with a grain of salt, since I haven't had time to polish it.

jonatack commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,356 @@
+Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-xxxx
+Status: Draft
+Type: Specification
+Created: 2025-08-22
```
```suggestion
Assigned: ?
```

jonatack commented on bip-encrypted-backup.md:
````diff
@@ -0,0 +1,356 @@
+```
+BIP: ?
+Title: Compact encryption scheme for non-seed wallet data
+Author: Pyth <pythcoiner@wizardsardine.com>
````
```suggestion
Authors: Pyth <pythcoiner@wizardsardine.com>
```

jonatack commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,356 @@
+Type: Specification
+Created: 2025-08-22
+License: BSD-2-Clause
+Post-History: https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607/31
```
```suggestion
Discussion: https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607/31
Requires: <BIP number(s)>
```

jonatack reviewed (commented):
A few updates for BIP 3.

jonatack commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,356 @@
+Created: 2025-08-22
+License: BSD-2-Clause
+Post-History: https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607/31
+              https://groups.google.com/g/bitcoindev/c/5NgJbpVDgEc
```
nit, link to the top post
```suggestion
              https://groups.google.com/g/bitcoindev/c/5NgJbpVDgEc/m/TtGK9sF9BgAJ
```

pythcoiner commented:
@jonatack thanks for the review, addressed your comments

pythcoiner commented:
I've done some [work](https://github.com/pythcoiner/bitcoin/tree/wip-encrypted-backup) on top of @Sjors implem & updated the rust [implem](https://github.com/pythcoiner/bitcoin-encrypted-backup), both have tests passing on the same set of test vectors and follows the actual version of this spec

murchandamus commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,356 @@
+BIP: ?
+Title: Compact encryption scheme for non-seed wallet data
+Authors: Pyth <pythcoiner@wizardsardine.com>
+Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-xxxx
```
The Comments-URI header is no longer used.
```suggestion
```

murchandamus commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,356 @@
+   depending on (often opaque) implementation details. Its security also reduces to
+   that of the weakest device with cloud access. Each copy increases the attack surface.
+
+This BIP therefore proposes an **encrypted**, and compact backup format that:
```
Stumbled on the "and" here:

```suggestion
This BIP therefore proposes an **encrypted**, compact backup format that:
```

murchandamus commented on bip-encrypted-backup.md:
```diff
@@ -0,0 +1,356 @@
+
+This BIP therefore proposes an **encrypted**, and compact backup format that:
+
+* can be **safely stored in multiple places**, including untrusted on-line services,  
```
```suggestion
* can be **safely stored in multiple places**, including untrusted online services,  
```

murchandamus commented on bip-0138.md:
```diff
@@ -0,0 +1,356 @@
+
+### Secret generation
+
+- Let $p_1, p_2, \dots, p_n$, be the public keys in the descriptor/wallet policy, in
```
It’s not clear to me what p₁, p₂, etc refer to here. Are they meant as the first few keys derived from the secret, or is this scheme only intended for descriptors/wallet policies with multiple keys? This should perhaps be explained a little further.

pythcoiner replied:
I've try to make it more clear.

murchandamus commented on bip-0138.md:
```diff
@@ -0,0 +1,356 @@
+  increasing lexicographical order
+- Let $s$ = sha256(sha256("BIPXXX_DECRYPTION_SECRET") | sha256("BIPXXX_DECRYPTION_SECRET") | $p_1$ | $p_2$ | ... | $p_n$)
+- Let $s_i$ = sha256(sha256("BIPXXX_INDIVIDUAL_SECRET") | sha256("BIPXXX_INDIVIDUAL_SECRET") | $p_i$)
+- Let $c_i$ = $s$ ⊕  $s_i$
```
It seems to me that for a single-sig policy this would always result in $c_i$ = 0x0…0.

If single-sig wallets are not eligible for this scheme, I must have missed where this was stated.

pythcoiner replied:
single sig descriptor are elligible to this format as `sha256(sha256("BIPXXX_DECRYPTION_SECRET") | sha256("BIPXXX_DECRYPTION_SECRET") | $p_1$ )`!= `sha256(sha256("BIPXXX_INDIVIDUAL_SECRET") | sha256("BIPXXX_INDIVIDUAL_SECRET") | $p_1$ )` as the tag differ, but maybe I miss something?

murchandamus replied:
No, I missed that! I’m gonna take another look at your recent updates and then see whether that was just me reading too quickly, or whether it could perhaps be stated more explicitly why the scheme is applicable to any set of public keys.

murchandamus commented on bip-encrypted-backup.md:
````diff
@@ -0,0 +1,356 @@
+```
+BIP: ?
````
```suggestion
BIP: ?
Layer: Applications
```

murchandamus commented on bip-0138.md:
```diff
@@ -0,0 +1,356 @@
+
+* **Encrypted**: safe to store with untrusted cloud providers or backup services  
+* **Access controlled**: only designated cosigners can decrypt  
+* **Easy to implement**: it should not require any sophisticated tools/libraries.  
```
I’m not sure that I correctly understood which public keys are being used to encrypt the data, but if it is the zeroth keys in the derivation sequence, wouldn’t any UTXO received to the zeroth address and later spent give a surveillant the information to decrypt the backup?

pythcoiner replied:
Good catch, you raise an important point, we must disallow key expressions w/o a trailing derivation!

murchandamus reviewed (commented):
I’m afraid that I was not able to complete follow your specification. Could you please describe in more detail how the keys are extracted from descriptor/wallet policy?
It’s also not obvious to me how initial use of the descriptor would not reveal the decryption secret to third parties.

The document is missing a Backwards Compatibility section.

pythcoiner commented:
Thanks @murchandamus for the review, I addressed your comments and reworked the spec:
- Better specification of which key is selected
- Disallowed keys that dont have a trailing derivation
- Updated test vectors

murchandamus committed:
Assign BIP138

murchandamus commented:
Assigned BIP138

murchandamus commented on bip-0138.md:
````diff
@@ -0,0 +1,373 @@
+```
+  BIP: 138
+  Layer: Applications
+  Title: Compact encryption scheme for non-seed wallet data
````
Nit: Personally, I prefer the titles of BIPs to be [title case](https://en.wikipedia.org/wiki/Title_case) as in the PR name, please feel free to disregard if you feel otherwise.

murchandamus commented on bip-0138.md:
```diff
@@ -0,0 +1,373 @@
+
+### Motivation
+
+Losing the **wallet descriptor** (or **wallet policy**) is just as catastrophic as
```
Losing the funds and privacy vs losing privacy does not seem equivalent. ;)

```suggestion
Losing the **wallet descriptor** (or **wallet policy**) is almost as catastrophic as
```

murchandamus commented on bip-0138.md:
```diff
@@ -0,0 +1,373 @@
+
+This proposal targets output script descriptors (BIP-0380) and policies (BIP-0388), but the
+scheme also works for labels (BIP-0329) and other wallet metadata like
+[wallet backup metadata](https://github.com/bitcoin/bips/pull/2130).
```
This should now perhaps be replaced by a link to BIP139

murchandamus reviewed (commented):
I just skimmed the proposal another time, this time without diving into the details.
Generally, this proposal is starting to get fairly mature. 

I’d like to invite some people that commented in the Delving discussion or are working on similar topics to give it another review. E.g., @bigspider, @Sjors, @reardencode, @craigraw, @seedhammer, @achow101, and @darosior, but please feel free to ignore if you don’t have time.

I notice that there isn’t a dedicated backwards compatibility section and would like to ask for one to be added. Please feel free to squash my number assignment commit into your commit (no need to give me credit).

What is your perspective on this submission, do you still have planned work, do you have some ideas for reviewers that should take another look before it is published?

bigspider commented on bip-0138.md:
```diff
@@ -0,0 +1,373 @@
+
+## Specification
+
+Note: in the followings sections, the operator ⊕  refers to the bitwise XOR operation.
```
Extra space
```suggestion
Note: in the followings sections, the operator ⊕ refers to the bitwise XOR operation.
```

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,373 @@
+
+#### Content
+
+`CONTENT` is a variable length field defining the type of `PLAINTEXT` being encrypted,
```
I find the current BIP text unclear about how multiple data blobs are supposed to be encoded. The term `CONTENT` is also too ambiguous.

Sjors replied:
Something like this would be better I think:

```diff
diff --git a/bip-0138.md b/bip-0138.md
index 2fbb2b2..1ed60bd 100644
--- a/bip-0138.md
+++ b/bip-0138.md
@@ -264,5 +264,5 @@ Implementations MUST reject empty payloads.
 defined in `ENCRYPTION` where `PAYLOAD` is encoded following this format:

-`CONTENT` `PLAINTEXT`
+`CONTENT` `CONTENT` `...`

 #### Integer Encodings
@@ -273,8 +273,8 @@ All variable-length integers are encoded as
 #### Content

-`CONTENT` is a variable length field defining the type of `PLAINTEXT` being encrypted,
+`CONTENT` is a variable length field containing one piece of encrypted content,
 it follows this format:

-`TYPE` (`LENGTH`) `DATA`
+`TYPE` `LENGTH` `DATA`

 `TYPE`: 1-byte unsigned integer identifying how to interpret `DATA`.
@@ -288,16 +288,13 @@ it follows this format:
 `LENGTH`: variable-length integer representing the length of `DATA` in bytes.

-For all `TYPE` values except `0x01`, `LENGTH` MUST be present.
-
 `DATA`: variable-length field whose encoding depends on `TYPE`.

 For `TYPE` values defined above:
 - 0x00: parsers MUST reject the payload.
-- 0x01: `LENGTH` MUST be omitted and `DATA` is a 2-byte big-endian unsigned integer
-  representing the BIP number that defines it.
+- 0x01: `DATA` MUST be at least 2 bytes. Its first two bytes are a big-endian unsigned
+  integer representing the BIP number that defines the remaining `DATA` bytes.
 - 0x02: `DATA` MUST be `LENGTH` bytes of opaque, vendor-specific data.

-For all `TYPE` values except `0x01`, parsers MUST reject `CONTENT` if `LENGTH` exceeds
-the remaining payload bytes.
+Parsers MUST reject `CONTENT` if `LENGTH` exceeds the remaining payload bytes.

 Parsers MUST skip unknown `TYPE` values less than `0x80`, by consuming `LENGTH` bytes
@@ -305,5 +302,5 @@ of `DATA`.

 For unknown `TYPE` values greater than or equal to `0x80`, parsers MUST stop parsing
-`CONTENT`.[^type-upgrade]
+content items.[^type-upgrade]

 [^type-upgrade]: **Why the 0x80 threshold?**
@@ -314,6 +311,7 @@ For unknown `TYPE` values greater than or equal to `0x80`, parsers MUST stop par
 #### BIP380 Descriptor Backup Content

-When `CONTENT` is `TYPE = 0x01` with `DATA = 0x017c` (BIP380), `PLAINTEXT`
-is UTF-8 encoded BIP380 descriptor backup content. It is either:
+When `CONTENT` is `TYPE = 0x01` and `DATA` begins with `0x017c` (BIP380),
+the remaining `DATA` bytes are UTF-8 encoded BIP380 descriptor backup content.
+It is either:

 - A text descriptor backup document.
```

Sjors commented on bip-0138.md:
Claude suggested the following to make padding work with multiple items:

```diff
 | Value  | Definition                             |
 |:-------|:---------------------------------------|
-| 0x00   | Reserved                               |
+| 0x00   | End of content items; padding follows  |
 | 0x01   | BIP Number (big-endian uint16)         |
 | 0x02   | Vendor-Specific Opaque Tag             |
```

(and elsewhere don't say `0x00: parsers MUST reject the payload.`)

pythcoiner replied:
it's also how I've implemented it, I'll update the spec

Sjors replied:
Let me know when that's done and I'll give it another try on the implementation side.

pythcoiner commented:
and on the same topic, as derivation paths are optionals maybe we should state in the spec that for decryption, if common derivation path are not present in the list, the implem should iterate over the list of derivation paths for account 0-10.
@Sjors would you implement such a thing?

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+`CONTENT_TYPE` is a variable length field defining the type of the `PLAINTEXT` that follows
+it. It follows this format:
+
+`TYPE` (`LENGTH`) `DATA`
```
Using TLV (type, length, data) here is confusing. At first glance the reader might think DATA is the actual content, but it's not.

Maybe use: `TYPE `, (`TYPE_LENGTH`), `TYPE_PARAMS`.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+`CIPHERTEXT` is the encrypted data resulting from encryption of `PAYLOAD` with algorithm
+defined in `ENCRYPTION` where `PAYLOAD` is encoded following this format:
+
+`CONTENT_TYPE` `LENGTH` `PLAINTEXT` (`CONTENT_TYPE` `LENGTH` `PLAINTEXT` ...) (`PADDING`)
```
Maybe rename `LENGTH` to `CONTENT_LENGTH` and `PLAINTEXT` to `CONTENT`.

Sjors replied:
The recent change from having 1 content type to multiple, makes the current use of `PLAINTEXT` incorrect, so I think it's fine to drop it. As my LLM puts it:

```
CIPHERTEXT                       ← encrypted bytes
  └ PAYLOAD (the plaintext)      ← what you decrypt
      ├ CONTENT_TYPE  CONTENT_LENGTH  CONTENT   ← item 1
      │   └ CONTENT_TYPE = TYPE [TYPE_LENGTH] TYPE_PARAMS
      ├ … (more items)
      └ PADDING
```

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+  Assigned: 2026-05-11
+  License: BSD-2-Clause
+  Discussion: https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607
+              https://groups.google.com/g/bitcoindev/c/5NgJbpVDgEc/m/TtGK9sF9BgAJ
```
Maybe add this pull request, since a lot of the design was discussed here. BIP texts are published on multiple websites, where the link to the original PR is not as easy to find as here on Github.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+### Abstract
+
+This BIP defines a compact encryption scheme for **output script descriptors** (BIP-0380),
+**wallet policies** (BIP-0388), **labels** (BIP-0329), and **wallet backup metadata** (BIP-0139).
```
Maybe replace this line with "and other wallet metadata", since the Scope section already enumerates these. It also avoids the impression that this BIP is strictly limited to the items listed here.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+
+Losing the **wallet descriptor** (or **wallet policy**) is almost as catastrophic as
+losing the seed itself. The seed lets you sign, but the descriptor maps you to your coins.
+For multisig or miniscript wallets, keys alone won't help: without the descriptor, you
```
"won't help" -> "are insufficient" or "don't cut it"

(they do help, but you need more)

And then maybe add:

> Multisig setups can be designed to retain access to coins even if one of the seeds is lost, for example a 3-of-3 MuSig2 keypath with a script path that lets 2-of-3 recover. But when one seed is lost, its public key may also be lost, rendering the coins inaccessible. This risk is not intuitive.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+
+1. **Descriptors are hard to store offline.**  
+   Descriptors can be much longer than a 12/24-word seed. Paper and steel backups
+   become impractical or error-prone.
```
Maybe add:

> While USB sticks and other offline digital storage solutions degrade over time and are unlikely to survive flood or fire.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+   become impractical or error-prone.
+
+2. **Online redundancy carries privacy risk.**  
+   USB drives, phones, and cloud storage solve the length problem but expose your
```
If you take my suggestion above, you should drop "USB drive" from the list here, since it's offline.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+* can be **safely stored in multiple places**, including untrusted online services,  
+* can be **decrypted only by intended holders** of specified public keys,
+
+See the original [Delving post](https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607/31)
```
Seems fine to keep this sentence, but it's already linked to in the Discussions section.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+This BIP therefore proposes an **encrypted**, compact backup format that:
+
+* can be **safely stored in multiple places**, including untrusted online services,  
+* can be **decrypted only by intended holders** of specified public keys,
```
> of specified _public_ keys

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+### Expected properties
+
+* **Encrypted**: safe to store with untrusted cloud providers or backup services  
+* **Access controlled**: only designated cosigners can decrypt  
```
Maybe split this, to also make the trade-off more clear: 

> **Access controlled**
> - only designated cosigners can decrypt, using their xpub
> - can't be decrypted using public keys revealed on chain

And see below.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+
+* **Encrypted**: safe to store with untrusted cloud providers or backup services  
+* **Access controlled**: only designated cosigners can decrypt  
+* **Easy to implement**: it should not require any sophisticated tools/libraries.  
```
(continued from above)

Let's also split the implementation section to hint at why we're using xpub rather than the more secure xprv for this:

> **Easy to implement**
> - it should not require any sophisticated tools/libraries
> - it should not require access to private key material (which secure signing systems may not yield)

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+scheme also works for labels (BIP-0329) and other wallet metadata like
+wallet backup metadata (BIP-0139).
+
+Private key material MUST be removed before encrypting any payload.
```
> This is because the use of _public key_ material for decryption makes the scheme unsuitable for storing privacy key material.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+  in particular it supports single-signature descriptors ($n = 1$). Each $p_i$
+  is the x-only-normalized root public key of one *allowed* key expression
+  (see [Descriptor key requirements](#descriptor-key-requirements)).
+- Let $s$ = sha256(sha256("BIP138_DECRYPTION_SECRET") | sha256("BIP138_DECRYPTION_SECRET") | $p_1$ | $p_2$ | ... | $p_n$)
```
The repeated `BIP138_DECRYPTION_SECRET` makes this hard to read. Let's use a notation for tagged hashes and reference to BIP340:

> - Let $s$ = tagged_hash("BIP138_DECRYPTION_SECRET", $p_1$ | $p_2$ | ... | $p_n$)
> - Let $s_i$ = tagged_hash("BIP138_INDIVIDUAL_SECRET", $p_i$)
> - Let $c_i$ = $s$ ⊕ $s_i$
>
> where tagged_hash(tag, m) = sha256(sha256(tag) | sha256(tag) | m) is the
> tagged hash construction from [BIP340](bip-0340.mediawiki#design).
```

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+- Ledger: `m/87h/<coin>h/{0..9}h`
+
+If fetching an xpub for one of these common paths fails or times out,
+implementations SHOULD warn the user.
```
> and description fails, implementations SHOULD warn the user.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,576 @@
+
+Encoders SHOULD append random decoy `INDIVIDUAL_SECRET` entries to hide the exact
+number of real decryption keys. The total count SHOULD be rounded up to the
+smallest bucket greater than or equal to the real count, starting at 5 and doubling
```
I guess 5 is fine as the minimum size.

Take a simple 3-of-3 musig2 with a single 2-of-3 fallback keypath, stored with with a plain text BIP388 policy.

Without the decoys the backup would be 844 base64 characters, with it this grows to 932, so only 10% extra.

xpubs make for the bulk of the payload size, and we really need those.

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,592 @@
+
+Padding is OPTIONAL and is an encoder concern only: decoders ignore the `PADDING` bytes and
+MUST NOT validate any bucket. Because the padding sits inside `CIPHERTEXT`, the
+`ENCRYPTED_PAYLOAD` `LENGTH` reveals only the padded size, not the real `CONTENT` size.
```
`CONTENT` -> `PAYLOAD`

kloaec commented on bip-0138.md:
```diff
@@ -0,0 +1,600 @@
+
+Applications that exclude additional keys SHOULD document this, although decryption
+using these keys will simply fail. This does not affect decryption with the remaining
+keys.
```
Unclear phrasing without contextualizing heavily.
Replacing the many "This", "that", "these" with what they actually are would be helpful.

murchandamus replied:
In technical writing, please feel free to reuse the same term repeatedly as you are going for clarity, not prose. :)

pythcoiner replied:
done

kloaec commented on bip-0138.md:
```diff
@@ -0,0 +1,600 @@
+keys, explicitly exclude any public keys with x coordinate
+`50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0` (the BIP341 NUMS
+point, used as a taproot internal key in some applications). Additionally, exclude any
+other publicly known keys.
```
Define "publicly known keys". Is there to be a registry of such keys?

Sjors replied:
All keys published on the blockchain, but I don't think there's a database elsewhere.

One thing that could be enumerated is to discourage BIP44/49/84/etc account level xpubs, because they may have been shared with a (hardware) wallet backend. That's dangerous (for privacy) if you keep one of your backups with a cloud provider.

pythcoiner replied:
> (hardware) wallet backend

@Sjors I'm not sure what you mean by this

Sjors replied:
E.g. the Ledger and Trezor desktop apps get your balance and transaction history from a server, which might know your xpub.

pythcoiner replied:
ok, was not aware they have this design
but I think "discouraging" this path will just have no effect
should we consider this BIP is not compatible with single-sig descriptors instead?

Sjors replied:
It's not that useful for single-sig descriptors, but I'm not sure if it's worth banning. An adversary who knows the xpub already knows most of the interesting stuff in the backup.

What's more important is that people don't use single-sig xpubs, previously used with such a wallet server backend, in their multisig setup. Because that could compromise their descriptor.

pythcoiner commented:
@kloaec reworded to make more explicit what could be a known key and added a warning on the point mentioned by @Sjors

Sjors commented on bip-0138.md:
```diff
@@ -0,0 +1,613 @@
+that xpub. For a single-signature descriptor this leaks little the server does not
+already know, but reusing such an xpub as a key of a multisig descriptor lets that
+server decrypt every backup of the multisig. Users SHOULD build a multisig from
+accounts whose xpub was never sent to such a server.
```
, such as BIP87.

pythcoiner commented:
@Sjors applied you suggestion (and added BIP48)

baslabofhd777-lab commented:
كيف

bigspider commented on bip-0138.md:
```diff
@@ -0,0 +1,613 @@
+[^psbt-base64]: **Why Base64?**
+    PSBT (BIP174) is commonly exchanged as a Base64 string, so wallet software
+    likely already supports this representation.
+
```
As some software wallets might want to save the encrypted backup to a file (for example, in order to upload to cloud storage) - perhaps it would be useful to recommend a standard file extension?

pythcoiner replied:
i'd just use `.bip138`

Sjors replied:
Wouldn't it be better to use binary encoding for such a file, and add magic bytes? With base64 I would just use .txt so it's easier to manually open and copy-paste if needed.

pythcoiner replied:
> Wouldn't it be better to use binary encoding for such a file, and add magic bytes?

I'm not sure to follow, this spec already define a binary encoding, and starts by `BIP138` magic bytes

see the output of a backup generated by the rust cli:

```
~/ws/bip138 master ?6 ❯ cat 3.bip138     
BIP138�0��
�b}~���g�
         l��-8��մj҄�^-�h���nd`*��)�      �5���͈F
                                              �B� ������r�!     ����B�U�t_�>�2�^�iQ���  ��z
                                                                                           S�����@�$�bM�c�
�w��V��*�6���c��a������e���*�OC�P.��Yq~��ic8�����8���H��'c�6����8(��}�����&X<�                            B��[KV}סu2՟�ddfzeŐA�v��
▒n0Ӄ�c�"B<F�a H����▒�]�6�A��                                           ���"▒��e�
                            넷H
��l�(��"_�J��Yfzr�,=���pB��eT���.bQ>[\▒���b�P��a�L<ȩl�\+���E�gX�vj�;X
                                                                     �-�b8?O��m��d����C:�����<�CJ�^�����|��?i��#������Xd�
g,���
������f�63Y�^�cs� XYO��-�0��;��YSQ�ĝ��z�_�t�|�~���
                                                  ��ݗ��$\�d�
/o���~�γ���M�K�L�G�=������>(���T�nRh��?|�:�f��f^'�%�ˋ�j\��%e�rC��cT�ҫ�q�\�w��q���H�a�DC���~C��R),y&u��,�|�r.▒�~�M3�~��g�=�/�d�4`�Up�R��~ϋ*�%�p�:�<�WԄKn���mS�E1��`n�N���U�@6��"���
              1b����x��|^�j-�
                             �&���M���/=�����6���o-�}�J�΄��ę,��L�&��5b��y�[�c<м▒�@�'�S  �r�:Bd�ĺ�b.Jp��.�1[�"!~3�4Ӵ���N]Px��W���16E��F�YH�L��d!�m�g����9)����i�y��jQh�K�C������i�t��j�▒�� Z&���l����t�|▒y#U�e
```

Sjors replied:
Ah I forgot. Ok, so then `.bip138` makes sense as an extension.

pythcoiner replied:
added a note for both formats

Sjors commented on bip-0138.md:
```diff
@@ -556,6 +556,12 @@ Implementations SHOULD encode and decode the backup using Base64 (RFC 4648).[^ps
+### File Extension
+
+Backups saved to a file SHOULD use the `.bip138` file extension, whether the file
+contains the raw binary encoding or its Base64 text representation. A Base64 backup
```
> or its Base64 text representation

I think .txt is better in base64. It's easier to open as text, and it also removes the need for implemtatinons to check whether a .bip138 file is binary or text encoded.

pythcoiner replied:
my bad, done
